scikeras/wrappers.py

```diff
@@ -888,84 +876,64 @@ def preprocess_y(
-                        "classes_": encoder[1].categories_[0],
-                        "n_classes_": encoder[1].categories_[0].size,
-                        "target_encoder_": encoder,
-                        "model_n_outputs_": 1,
-                        "n_outputs_": 1,
-                    }
-                )
-            else:
-                encoder = self.target_encoder_
-            y = encoder.transform(y)
-        elif self.target_type_ == "multiclass":
-            # y = array([1, 5, 2])
-            # encode
-            if reset:
-                if is_categorical_crossentropy(loss):
-                    # one-hot encode
-                    encoder = make_pipeline(
-                        Ensure2DTransformer(),
-                        OneHotEncoder(sparse=False, dtype=np.float32),
-                    )
-                else:
-                    encoder = make_pipeline(
-                        Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
-                    )
-                y = encoder.fit_transform(y)
-                meta.update(
-                    {
-                        "classes_": encoder[1].categories_[0],
-                        "n_classes_": encoder[1].categories_[0].size,
-                        "target_encoder_": encoder,
-                        "model_n_outputs_": 1,
-                        "n_outputs_": 1,
-                    }
-                )
-            else:
-                encoder = self.target_encoder_
-                y = encoder.transform(y)
+        encoders = {
+            "binary": make_pipeline(
+                Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
+            ),
+            "multiclass": make_pipeline(
+                Ensure2DTransformer(), OrdinalEncoder(dtype=np.float32),
+            ),
+            "multiclass-one-hot": FunctionTransformer(),
+        }
+        if is_categorical_crossentropy(loss):
+            encoders["multiclass"] = make_pipeline(
+                Ensure2DTransformer(), OneHotEncoder(sparse=False, dtype=np.float32),
+            )
+
+        if self.target_type_ not in encoders:
+            raise ValueError(
+                f"Unknown label type: {self.target_type_}."
+                "\n\nTo implement support, subclass KerasClassifier and override"
+                " `preprocess_y` and `postprocess_y`."
+                "\n\nSee (TODO: link to docs) for more information."
+            )
+
+        if not hasattr(self, "target_encoder_"):
+            self.target_encoder_ = encoders[self.target_type_].fit(y)
+
+        y = self.target_encoder_.transform(y)
+        encoder = self.target_encoder_
+
+        if self.target_type_ in ["binary", "multiclass"]:
+            meta.update(
+                {
+                    "classes_": encoder[1].categories_[0],
+                    "n_classes_": encoder[1].categories_[0].size,
+                }
+            )
         elif self.target_type_ == "multiclass-one-hot":
```

Comment:
We end up still keeping some `if/elif/else` logic here. We also have the same logic in `postprocess_y`. While I think this is cleaner, I like that in the old implementation there was a 1:1 map between the logic in `preprocess_y` and `postprocess_y`. Not a big deal, it's still clean, but I just wanted to point it out.